packages/itwinui-react/src/core/Dialog/DialogMain.tsx

```diff
@@ -203,36 +202,53 @@ export const DialogMain = React.forwardRef((props, ref) => {
     </Box>
   );
 
-  return (
-    <Transition
-      in={isOpen}
-      appear={true}
-      timeout={{ exit: 600 }}
-      // Focuses dialog when opened
-      onEntered={() => {
-        previousFocusedElement.current = dialogRef.current?.ownerDocument
-          .activeElement as HTMLElement;
-        setFocus && dialogRef.current?.focus({ preventScroll: true });
-      }}
-      // Brings back focus to the previously focused element when closed
-      onExit={() => {
-        if (
-          dialogRef.current?.contains(
-            dialogRef.current?.ownerDocument.activeElement,
-          )
-        ) {
-          previousFocusedElement.current?.focus();
-        }
-      }}
-      unmountOnExit={true}
-      nodeRef={dialogRef}
-    >
-      <DialogDragContext.Provider value={{ onPointerDown: handlePointerDown }}>
-        {trapFocus && <FocusTrap>{content}</FocusTrap>}
-        {!trapFocus && content}
-      </DialogDragContext.Provider>
-    </Transition>
-  );
+  /** Focuses dialog when opened. */
+  const onEnter = React.useCallback(() => {
+    previousFocusedElement.current = dialogRef.current?.ownerDocument
+      .activeElement as HTMLElement;
+    setFocus && dialogRef.current?.focus({ preventScroll: true });
+  }, [setFocus]);
+
+  /** Brings back focus to the previously focused element when closed. */
+  const onExit = React.useCallback(() => {
+    if (
+      dialogRef.current?.contains(
+        dialogRef.current?.ownerDocument.activeElement,
+      )
+    ) {
+      previousFocusedElement.current?.focus();
+    }
+  }, []);
+
+  const [prevIsOpen, setPrevIsOpen] = React.useState<typeof isOpen>(false);
+  const [shouldBeMounted, setShouldBeMounted] = React.useState(isOpen);
+
+  if (prevIsOpen !== isOpen) {
+    setPrevIsOpen(isOpen);
+
+    if (isOpen) {
+      // Mount *before* handling dialog entry.
+      queueMicrotask(() => {
+        onEnter();
+      });
+
+      setShouldBeMounted(true);
+    } else {
+      onExit();
+
+      // Wait for the animation to end
+      setTimeout(() => {
+        setShouldBeMounted(false);
+      }, 600);
+    }
+  }
```

Comment:
Can we use a callback ref to get precise timing of when the element mounts/unmounts?

e.g. something like this:

```jsx
ref={
  React.useCallback((element) => {
    if (element) {
      onEnter();
    } else {
      onExit();
    }
  }, [])
}
```